node/core/provisioner/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+		match from_job {
+			FromJob::Runtime(ram) => AllMessages::RuntimeApi(ram),
+		}
+	}
+}
+
+impl TryFrom<AllMessages> for FromJob {
+	type Error = ();
+
+	fn try_from(msg: AllMessages) -> Result<Self, Self::Error> {
+		match msg {
+			AllMessages::RuntimeApi(runtime) => Ok(FromJob::Runtime(runtime)),
+			_ => Err(()),
+		}
+	}
+}
+
+#[derive(Debug, derive_more::From)]
+pub enum Error {
+	#[from]
+	Sending(mpsc::SendError),
+	#[from]
+	Util(util::Error),
+	#[from]
+	OneshotRecv(oneshot::Canceled),
+	OneshotSend,
+}
+
+impl JobTrait for ProvisioningJob {
+	type ToJob = ToJob;
+	type FromJob = FromJob;
+	type Error = Error;
+	type RunArgs = ();
+
+	const NAME: &'static str = "ProvisioningJob";
+
+	/// Run a job for the parent block indicated
+	//
+	// this function is in charge of creating and executing the job's main loop
+	fn run(
+		relay_parent: Hash,
+		_run_args: Self::RunArgs,
+		receiver: mpsc::Receiver<ToJob>,
+		sender: mpsc::Sender<FromJob>,
+	) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>> {
+		async move {
+			let job = ProvisioningJob::new(relay_parent, sender, receiver);
+
+			// it isn't necessary to break run_loop into its own function,
+			// but it's convenient to separate the concerns in this way
+			job.run_loop().await
+		}
+		.boxed()
+	}
+}
+
+impl ProvisioningJob {
+	pub fn new(
+		relay_parent: Hash,
+		sender: mpsc::Sender<FromJob>,
+		receiver: mpsc::Receiver<ToJob>,
+	) -> Self {
+		Self {
+			relay_parent,
+			sender,
+			receiver,
+			provisionable_data_channels: Vec::new(),
+			backed_candidates: Vec::new(),
+			signed_bitfields: Vec::new(),
+		}
+	}
+
+	async fn run_loop(mut self) -> Result<(), Error> {
+		while let Some(msg) = self.receiver.next().await {
+			use ProvisionerMessage::{
+				ProvisionableData, RequestBlockAuthorshipData, RequestInherentData,
+			};
+
+			match msg {
+				ToJob::Provisioner(RequestInherentData(_, sender)) => {
+					// Note the cloning here: we have to clone the vectors of signed bitfields and backed candidates
+					// so that we can respond to more than a single request for inherent data; we can't just move them.
+					// It would be legal, however, to set up `from_job` as `&mut _` instead of a clone. We clone it instead
+					// of borrowing it so that this async function doesn't depend at all on the lifetime of `&self`.
+					send_inherent_data(
+						self.relay_parent,
+						self.signed_bitfields.clone(),
+						self.backed_candidates.clone(),
+						sender,
+						self.sender.clone(),
+					)
+					.await?
+				}
+				ToJob::Provisioner(RequestBlockAuthorshipData(_, sender)) => {
+					self.provisionable_data_channels.push(sender)
+				}
+				ToJob::Provisioner(ProvisionableData(data)) => {
+					for channel in self.provisionable_data_channels.iter_mut() {
+						// REVIEW: the try operator here breaks the run loop if any receiver ever unexpectedly
+						// closes their channel. Is that desired?
+						channel.send(data.clone()).await?;
```

Comment:
What are these channels sending to?

Reply:
we don't want to break the run loop here, no. but i'm now questioning why this `RequestBlockAuthorshipData` even exists...

Reply:
Anyone at all can send a `RequestBlockAuthorshipData` to the provisioner; we don't know who originates them. We just ensure that we distribute the authorship data to each of them. 

As for why it exists, [the guide says](https://w3f.github.io/parachain-implementers-guide/node/utility/provisioner.html#block-production):

> One might ask: given `ProvisionerMessage::RequestInherentData`, what's the point of `ProvisionerMessage::RequestBlockAuthorshipData`? The answer is that the block authorship data includes more information than is present in the inherent data; disputes, for example.